### PR TITLE
added tests for caching and stale response

### PR DESCRIPTION
**Pull request by swaroopdantu**

### CODE REVIEW

**Comment by swaroopdantu** [comment]
Results :

Tests run: 3, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 27.716 s
[INFO] Finished at: 2018-10-03T13:51:12-07:00
[INFO] ------------------------------------------------------------------------

File: src/main/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptor.java [hunk 1]
```diff
@@ -16,16 +16,43 @@
...
 public class StaleIfErrorInterceptor implements Interceptor {
     // based on code from https://github.com/square/okhttp/issues/1083 by jvincek
+	private static final int DEFULAT_MAX_STALE_DURATION = 28;
+	private static final TimeUnit DEFULAT_MAX_STALE_TIMEUNIT = TimeUnit.DAYS;
+	
+	private final int staleDuration;
+	private final TimeUnit staleDurationTimeUnit;
+	
+    public StaleIfErrorInterceptor(int staleDuration, TimeUnit staleDurationTimeUnit) {
+		super();
```
Comment (inder123) [inline src/main/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptor.java]: should use empty super() in general, as it is assumed.

File: src/main/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptor.java [hunk 2]
```diff
@@ -16,16 +16,43 @@
...
+		this.staleDuration = staleDuration;
+		this.staleDurationTimeUnit = staleDurationTimeUnit;
+	}
+    
+    public StaleIfErrorInterceptor() {
+		super();
```
Comment (inder123) [inline src/main/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptor.java]: change this to:
this(DEFAULT_MAX_STALE_DURATION, TimeUnit.DAYS)

File: src/main/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptor.java [hunk 3]
```diff
@@ -43,14 +70,18 @@ public Response intercept(Interceptor.Chain chain) throws IOException {
...
 
-        if (response == null || !response.isSuccessful()) {
-            Request newRequest = request.newBuilder().cacheControl(CacheControl.FORCE_CACHE).build();
-            try {
-                response = chain.proceed(newRequest);
-            } catch (Exception e) { // cache not available
-                throw e;
-            }
-        }
+		if (response == null || !response.isSuccessful()) {
```
Comment (inder123) [inline src/main/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptor.java]: Can you fix the indentation?

File: src/test/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptorTest.java [hunk 4]
```diff
@@ -15,14 +15,310 @@
...
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+import okhttp3.tls.HandshakeCertificates;
+import static okhttp3.tls.internal.TlsUtil.localhost;
+
+/**
+ *
```
Comment (inder123) [inline src/test/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptorTest.java]: Please write a meaningful class comment as well

File: src/test/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptorTest.java [hunk 5]
```diff
@@ -15,14 +15,310 @@
...
+				.addHeader("Content-Type: text/html; charset=utf-8")
+				.addHeader("Cache-Control", "public, max-age=2");
+		server.enqueue(mockResponse);
+		server.enqueue(mock404Response);
+		server.enqueue(mockResponse);
+		
+		Request request = request().get().build();
+		Response response = null;
+		try {
+			//make first call and cache the response
```
Comment (inder123) [inline src/test/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptorTest.java]: space after //

File: src/test/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptorTest.java [hunk 6]
```diff
@@ -15,14 +15,310 @@
...
+			response = client.newCall(request).execute();
+			assertNotNull(response);
+			assertNotNull(response.cacheResponse());
+			assertNull(response.networkResponse());
+			assertEquals(200, response.code());
+			response.body().close();
+			try {
+				Thread.sleep(5000);
+			} catch (InterruptedException e) {
+				e.printStackTrace();
```
Comment (inder123) [inline src/test/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptorTest.java]: don't catch Thread.sleep, let it fail the test?

File: src/test/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptorTest.java [hunk 7]
```diff
@@ -15,14 +15,310 @@
...
+			
+			//bring back the server here
+			response = client.newCall(request).execute();
+			assertNotNull(response);
+			assertNull(response.cacheResponse());
+			assertNotNull(response.networkResponse());
+			response.body().close();
+			
+		} catch (IOException e) {
+			e.printStackTrace();
```
Comment (inder123) [inline src/test/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptorTest.java]: why do we catch exception?

File: src/test/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptorTest.java [hunk 8]
```diff
@@ -15,14 +15,310 @@
...
+		} finally {
+			if (response != null) {
+				response.body().close();
```
Comment (inder123) [inline src/test/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptorTest.java]: Can we use try-with-resources instead?

Comment (swaroopdantu) [reply src/test/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptorTest.java]: I think there is no need of entire try catch, will remove

File: src/test/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptorTest.java (hunk 7 above)
Comment (swaroopdantu) [reply src/test/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptorTest.java]: we dont need it, I will remove the try catch blocks.

File: src/test/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptorTest.java [hunk 9]
```diff
@@ -15,14 +15,310 @@
...
+	private OkHttpClient client;
+	private OkHttpClient clientWith4SecMaxStale;
+
+	private HttpUrl url;
+
+	@Before
+	public void setUp() {
+		client = new OkHttpClient.Builder()
+				.addInterceptor(new StaleIfErrorInterceptor())
+				.sslSocketFactory(handshakeCertificates.sslSocketFactory(), handshakeCertificates.trustManager())
```
Comment (inder123) [inline src/test/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptorTest.java]: Do we really need sslSocketFactory?

Comment (swaroopdantu) [reply src/test/java/com/peel/okhttp3/staleiferror/StaleIfErrorInterceptorTest.java]: we don't need it for our current tests, removed it.

**Pull request merged**